### PR TITLE
Do additional character escaping for wallet names and address labels

### PR DESCRIPTION
**Pull request by achow101**
Fixes some places where wallet names and address labels which contain valid html or other interpreted characters are displayed incorrectly.

In the send coins dialog, if the wallet name or the address label contains valid html, then the html would be shown rather than the literal string for the wallet name or label. This PR fixes that so the true name or label is shown.

The Open Wallet menu would incorrectly show wallet names with ampersands (`&`). For some reason, Qt removes the first ampersand in a string. So by replacing the first ampersand with 2 ampersands, the correct number of ampersands will be shown.

Fixes the HTML escaping issues in #16820

### CODE REVIEW

**Review by jonatack — COMMENTED** [review]
Tested ACK ba9f82a3ec81b8cd8fd729e43b87e93debbd7fba for the ampersand issue. I don't know how to test the send coins dialog issue. The code looks fine.

**Comment by darosior** [comment]
ACK ba9f82a3ec81b8cd8fd729e43b87e93debbd7fba

**Comment by kristapsk** [comment]
> For some reason, Qt removes the first ampersand in a string.

First ampersand is removed because it is used to specify hot key. Replacing it with "&&" is recommended method. https://doc.qt.io/qt-5/qmenubar.html#details

**Comment by achow101** [comment]
> I don't know how to test the send coins dialog issue.

Make a wallet named like `<b>A</b>`. It should appear as `<b>A</b>` everywhere. But in the send coins dialog, on master, it will appear as **A** (bold A). The send coins dialog is the confirmation dialog that appears after you click Send on the send page.

When you are on the send page, set a label for the recipient in the same way that the wallet name was set. Then check the send coins dialog.

**Review by kristapsk — APPROVED** [review]
ACK ba9f82a3ec81b8cd8fd729e43b87e93debbd7fba (tested both cases with and without these changes)

**Comment by jonatack** [comment]
Thanks @achow101.

**Comment by jonatack** [comment]
ACK ba9f82a3ec81b8cd8fd729e43b87e93debbd7fba, I have reviewed the code and manually tested these changes, and agree they can be merged.

Additional test to my previous manual testing, with send label supplied as `<b>A</b>`:

On master:
```
Are you sure you want to send?
Please, review your transaction.

0.00100000 BTC from wallet 'PR16826' to 'A'  (A in bold)
```
This PR:
```
Are you sure you want to send?
Please, review your transaction.

0.00100000 BTC from wallet 'PR16826' to '<b>A</b>'
```

**Comment by kristapsk** [comment]
There is similar issue with close wallet dialog, maybe could be fixed with this PR too.

image

**Comment by jonatack** [comment]
@kristapsk good catch, reproduced your issue.

**Comment by achow101** [comment]
Fixed the escaping in the close dialog. Also noticed that notifications needed HTML escaping too, so I've added escapes for the wallet name and address label in transaction notifications.

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -375,6 +375,8 @@ void BitcoinGUI::createActions()
             for (const std::pair<const std::string, bool>& i : m_wallet_controller->listWalletDir()) {
                 const std::string& path = i.first;
                 QString name = path.empty() ? QString("["+tr("default wallet")+"]") : QString::fromStdString(path);
+                // Menu items remove single &. Single & are shown when && is in the string, but only the first occurrence. So replace only the first & with &&
```
[the remark this answers is not in the document]
Comment (promag) [reply src/qt/bitcoingui.cpp]: Reference https://doc.qt.io/qt-5/qshortcut.html#mnemonic.

**Comment by kristapsk** [comment]
Tested close wallet dialog, is ok now.

File: src/qt/walletcontroller.cpp [hunk 2]
```diff
@@ -75,7 +75,7 @@ void WalletController::closeWallet(WalletModel* wallet_model, QWidget* parent)
 {
     QMessageBox box(parent);
     box.setWindowTitle(tr("Close wallet"));
-    box.setText(tr("Are you sure you wish to close wallet <i>%1</i>?").arg(wallet_model->getDisplayName()));
+    box.setText(tr("Are you sure you wish to close wallet <i>%1</i>?").arg(GUIUtil::HtmlEscape(wallet_model->getDisplayName())));
```
Comment (jonatack) [inline src/qt/walletcontroller.cpp]: Re DrahtBot: this PR and #16822 both touch this one line. Trivial rebase for the whichever is not merged first.

**Comment by laanwj** [comment]
Untested ACK, thanks for adding proper escaping, ad52f054f67374dc46e0096d1e2f593d6372a2df

**Review by fanquake — APPROVED** [review]
ACK ad52f054f67374dc46e0096d1e2f593d6372a2df

Ampersand escaping. [master](08bb4c31563505a30a9b7dd8e7ed98966d2c8494):
alice_bob_master

#16826:
alice_bob_16826

HTML escaping. [master](08bb4c31563505a30a9b7dd8e7ed98966d2c8494):
slanty_master

#16826
slanty_16826

File: src/qt/bitcoingui.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply src/qt/bitcoingui.cpp]: @jonatack Given you'll need to rebase #16822, you can fix the comment there if you want. I don't mind either way.

[the remark this answers is not in the document]
Comment (jonatack) [reply src/qt/bitcoingui.cpp]: Appended the change to https://github.com/bitcoin/bitcoin/pull/16822/commits/cad3ab5db835e1d0c44c6a5fc0e4b2d1661dcd5c while rebasing.

**Pull request merged**